### PR TITLE
wildcard tag search support

### LINKED ISSUE #937: allow wildcards when searching for tags
I am using prefixed tags (e.g. `customer-A, customer-b, lang-css, lang-js, ...`) and would like to use wildcards when searching for links. So if searched for `lang-*`,  I would find all tags (links with tags) that start with `lang-`.

Would you accept a PR that introduces asterisks as wildcard placeholders when searching for tags? The only implication I can think of is that searching for tags containing asterisks could possibly return more results than expected.

### PR DESCRIPTION
**Pull request by thewilli**
- when searching for tags you can now include '*' as wildcard placeholder
- new search reduces overall overhead when filtering for tags
- fixed combination with description tag search ('#' prefix)
- tests added

fixes #937

### CODE REVIEW

File: application/LinkFilter.php [hunk 1]
```diff
@@ -250,6 +250,50 @@ private function filterFulltext($searchterms, $visibility = 'all')
...
+        if($tag[0] === "-") {
+            // query is negated
+            $i = 1; // use offset to start after '-' character
+            $regex = '(?!'; // create negative lookahead
+        } else {
+            $i = 0; // start at first character
+            $regex = '(?='; // use positive lookahead
+        }
+        $regex .= '.*(?:^| )'; // before tag may only be a space or the beginning
+        $len = strlen($tag);
```
Comment (ArthurHoaro) [inline application/LinkFilter.php]: Very slight optimization: you call already `strlen` at the beginning of your method.

File: tests/api/controllers/GetLinksTest.php [hunk 2]
```diff
@@ -367,6 +367,30 @@ public function testGetLinksSearchTags()
...
+        $env = Environment::mock([
+            'REQUEST_METHOD' => 'GET',
+            'QUERY_STRING' => 'searchtags=w*b',
+        ]);
+        $request = Request::createFromEnvironment($env);
+        $response = $this->controller->getLinks($request, new Response());
+        $this->assertEquals(200, $response->getStatusCode());
+        $data = json_decode((string) $response->getBody(), true);
+        $this->assertEquals(4, count($data));
+        $this->assertEquals(6, $data[0]['id']);
```
Comment (ArthurHoaro) [inline tests/api/controllers/GetLinksTest.php]: A few more use cases would be nice here (especially multiple tags and exclusion).

File: application/LinkFilter.php (hunk 1 above)
Comment (thewilli) [reply application/LinkFilter.php]: @ArthurHoaro you're right, missed that, thank you. It's now fixed 😃

File: tests/api/controllers/GetLinksTest.php (hunk 2 above)
Comment (thewilli) [reply tests/api/controllers/GetLinksTest.php]: @ArthurHoaro  you're right, I've now added some more. It's pretty hard to construct useful ones without tampering your test data and breaking the other tests :wink:

Does this suffice for you, or can you think of any more you'd like to have?

Comment (ArthurHoaro) [reply tests/api/controllers/GetLinksTest.php]: It is, but rewriting a datastore anytime you want to add a new test would be even more painful.

Yes, it's fine!

**Pull request merged**